recipes-robot/servohack/servohack/servohack.py

```diff
@@ -0,0 +1,75 @@
+#!/usr/bin/env python3
+import asyncio
+import logging
+from json import JSONDecodeError, loads
+from typing import Match
+
+from astoria.common.components import StateConsumer
+from astoria.common.code_status import CodeStatus
+from astoria.common.ipc import ProcessManagerMessage
+from pydantic import parse_obj_as
+
+LOGGER = logging.getLogger("servohack")
+
+logging.basicConfig(
+    level=logging.INFO,
+    format=f"%(asctime)s %(name)s %(levelname)s %(message)s",
+    datefmt="%Y-%m-%d %H:%M:%S",
+)
+
+loop = asyncio.get_event_loop()
+
+
+class ServoHackConsumer(StateConsumer):
+    """Restart the USB Hubs when usercode finishes."""
+
+    name = "servohack"
+
+    def _setup_logging(self, verbose: bool, *, welcome_message: bool = True) -> None:
+        """Suppress INFO messages from gmqtt."""
+        logging.getLogger("gmqtt").setLevel(logging.WARNING)
+
+    def _init(self) -> None:
+        """Initialise consumer."""
+        self._lock = asyncio.Lock()
+        self._code_started = False
+
+        self._mqtt.subscribe("astprocd", self._handle_astprocd_message)
+
+    async def _handle_astprocd_message(
+        self,
+        match: Match[str],
+        payload: str,
+    ) -> None:
+        """Handle astprocd status messages."""
+        try:
+            message = parse_obj_as(ProcessManagerMessage, loads(payload))
+            if message.status == ProcessManagerMessage.Status.RUNNING:
+                LOGGER.debug(f"Received process info: {message}")
+                async with self._lock:
+                    if message.code_status is CodeStatus.RUNNING:
+                        self._code_started = True
+                        LOGGER.info("Code has been started")
+                    elif message.code_status is not CodeStatus.RUNNING and self._code_started:
+                        LOGGER.info("USB removed, toggling USB lines")
+                        await asyncio.create_subprocess_exec(
+                            "/usr/sbin/uhubctl",
+                            "--loc", "1-1",
+                            "--action", "cycle",
+                            "--delay", "0.2",
```

Comment:
Can we have a comment explaining what this command is doing.

Reply:
e8467c5

Reply:
This looks to only cycle the USB 2 ports. May need a separate command for the USB 3 ports.